### PR TITLE
Ledger store

### PR DESCRIPTION
**Pull request by jenr24G**
Removes the Ledger from the Witness Tree and moves it to RocksDB

* Create `IndexerStore` and traits `BlockStore` and `LedgerStore`
* Use column families on `IndexerStore` to distinguish between ledgers and blocks keyed under the same state hash
* Update `IndexerState::best_ledger` to iterate through blocks and generate a ledger
* Update tests to reflect changes to `IndexerState` and `Branch` classes

<hr>

Asana Tasks referenced:
* https://app.asana.com/0/1203669416123378/1204820393125456/f

### CODE REVIEW

File: src/store.rs [hunk 1]
```diff
@@ -0,0 +1,172 @@
...
+        })
+    }
+
+    pub fn db_path(&self) -> &Path {
+        &self.db_path
+    }
+}
+
+impl BlockStore for IndexerStore {
+    fn add_block(&self, block: &crate::block::precomputed::PrecomputedBlock) -> anyhow::Result<()> {
```
Comment (Isaac-DeFrain) [inline src/store.rs:53]: Just for consistency, can we write these `crate::...` as `use` thingies at the top of the file?

File: src/store.rs [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+        let key = block.state_hash.as_bytes();
+        let value = bcs::to_bytes(&block)?;
+        self.database.put_cf(&cf_handle, key, value)?;
+        Ok(())
+    }
+
+    fn get_block(
+        &self,
+        state_hash: &crate::block::BlockHash,
+    ) -> anyhow::Result<Option<crate::block::precomputed::PrecomputedBlock>> {
```
Comment (Isaac-DeFrain) [inline src/store.rs:67]: Same as above

File: src/store.rs [hunk 3]
```diff
@@ -0,0 +1,172 @@
...
+        }
+        Ok(precomputed_block)
+    }
+}
+
+impl LedgerStore for IndexerStore {
+    fn add_ledger(
+        &self,
+        state_hash: &crate::block::BlockHash,
+        ledger: crate::state::ledger::Ledger,
```
Comment (Isaac-DeFrain) [inline src/store.rs:90]: Same

File: src/store.rs [hunk 4]
```diff
@@ -0,0 +1,172 @@
...
+        let key = state_hash.0.as_bytes();
+        let value = bcs::to_bytes(&ledger)?;
+        self.database.put_cf(&cf_handle, key, value)?;
+        Ok(())
+    }
+
+    fn get_ledger(
+        &self,
+        state_hash: &crate::block::BlockHash,
+    ) -> anyhow::Result<Option<crate::state::ledger::Ledger>> {
```
Comment (Isaac-DeFrain) [inline src/store.rs:105]: Same

File: src/store.rs [hunk 5]
```diff
@@ -0,0 +1,172 @@
...
+
+impl BlockStore for IndexerStore {
+    fn add_block(&self, block: &crate::block::precomputed::PrecomputedBlock) -> anyhow::Result<()> {
+        let cf_handle = self
+            .database
+            .cf_handle("blocks")
+            .expect("column family exists");
+        let key = block.state_hash.as_bytes();
+        let value = bcs::to_bytes(&block)?;
+        self.database.put_cf(&cf_handle, key, value)?;
```
Comment (Isaac-DeFrain) [inline src/store.rs:60]: So cool!

**Review by Isaac-DeFrain — CHANGES_REQUESTED** [review]
The ol' LGTM, just a few comments

File: src/store.rs [hunk 6]
```diff
@@ -0,0 +1,172 @@
+use std::path::{Path, PathBuf};
+
+use rocksdb::{ColumnFamilyDescriptor, DBWithThreadMode, MultiThreaded};
+
+use crate::{block::store::BlockStore, state::ledger::store::LedgerStore};
+
+#[derive(Debug)]
+pub struct IndexerStore {
+    db_path: PathBuf,
+    database: DBWithThreadMode<MultiThreaded>,
```
Comment (trevorbernard) [inline src/store.rs:10]: Are we having multiple producers to this database?

**Review by trevorbernard — CHANGES_REQUESTED** [review]

File: src/store.rs (hunk 6 above)
Comment (jenr24G) [reply src/store.rs:10]: One producer, multiple column families!

File: src/state/mod.rs [hunk 7]
```diff
@@ -330,44 +346,70 @@ impl IndexerState {
...
+                    state_hash = BlockHash::from_hashv1(
+                        precomputed_block.protocol_state.previous_state_hash,
+                    );
+                }
+                for ledger_diff in ledger_diffs {
+                    ledger.iter_mut().for_each(|ledger| {
+                        ledger
+                            .apply_diff(ledger_diff.clone())
+                            .expect("ledger diff application succeeds")
+                    });
```
[the remark this answers is not in the document]
Comment (Isaac-DeFrain) [reply src/state/mod.rs]: oh duh

**Review by Isaac-DeFrain — APPROVED** [review]

**Review by trevorbernard — APPROVED** [review]

**Pull request merged**